drv/transceivers-server/src/net_api.rs

```diff
@@ -0,0 +1,414 @@
+            }
+        }
+        Ok(())
+    }
+
+    fn read(
+        &mut self,
+        mem: MemoryRegion,
+        modules: ModuleId,
+        out: &mut [u8],
+    ) -> Result<(), Error> {
+        // Reading beyond a 128-byte page boundary will wrap around; return an
+        // error if that's going to happen.
+        let page_end = (mem.offset() as u32 / 128 + 1) * 128;
+        if mem.offset() as u32 + mem.len() as u32 > page_end {
+            // TODO use a more descriptive error name here?
+            return Err(Error::InvalidMemoryAccess {
+                offset: mem.offset(),
+                len: mem.len(),
+            });
+        }
+        // CMIS spec only guarantee an 8-byte read (?!)
+        // (see CMIS 5.0, 5.2.2.1)
+        if matches!(mem.upper_page(), UpperPage::Cmis(..)) && mem.len() > 8 {
+            // TODO: more specialized error
+            return Err(Error::RequestTooLarge);
+        }
+
+        self.select_page(*mem.upper_page(), modules)?;
+        self.transceivers
+            .setup_i2c_read(mem.offset(), mem.len(), get_mask(modules)?)
+            .map_err(|_e| Error::ReadFailed)?;
+        sleep_for(5); // TODO: wait for completion
+
+        let controller = match modules.fpga_id {
+            0 => FpgaController::Left,
+            1 => FpgaController::Right,
+            i => return Err(Error::InvalidFpga(i)),
+        };
+        for (port, out) in modules
+            .ports
+            .to_indices()
+            .zip(out.chunks_mut(mem.len() as usize))
+        {
+            self.transceivers
+                .get_i2c_read_buffer(PortLocation { controller, port }, out)
+                .map_err(|_e| Error::ReadFailed)?;
+        }
+        Ok(())
+    }
+
+    fn write(
+        &mut self,
+        mem: MemoryRegion,
+        modules: ModuleId,
+        data: &[u8],
+    ) -> Result<(), Error> {
+        // Writing beyond a 128-byte page boundary will wrap around; return an
+        // error if that's going to happen.
+        let page_end = (mem.offset() as u32 / 128 + 1) * 128;
+        if mem.offset() as u32 + mem.len() as u32 > page_end {
+            // TODO use a more descriptive error name here?
+            return Err(Error::InvalidMemoryAccess {
+                offset: mem.offset(),
+                len: mem.len(),
+            });
+        }
+        match *mem.upper_page() {
+            // CMIS 5.0, §5.2.2.2: only guarantee an 8-byte write
+            UpperPage::Cmis(..) => {
+                if mem.len() > 8 {
+                    // TODO: more specialized error
+                    return Err(Error::RequestTooLarge);
+                }
+            }
+            // SFF-8636, §5.3.3: max of 4 bytes in one write transaction
+            UpperPage::Sff8636(..) => {
+                if mem.len() > 4 {
+                    // TODO: more specialized error
+                    return Err(Error::RequestTooLarge);
+                }
+            }
+        }
```

Comment:
Actually, _all_ of these checks should be obviated by the type system (once the PR mentioned above is merged). It won't be possible to construct a read for greater than 128 bytes (SFF-8636) or 8 bytes (CMIS), or a read greater than 4 (SFF) or 8 bytes (CMIS).

Reply:
Yup, fixed post-rebase